lib/auth/auth.go

```diff
@@ -468,6 +475,42 @@ func (a *Server) GenerateUserTestCerts(key []byte, username string, ttl time.Dur
 	return certs.ssh, certs.tls, nil
 }
 
+// GenerateUserAppTestCert generates an application specific certificate, used
+// internally for tests.
+func (a *Server) GenerateUserAppTestCert(publicKey []byte, username string, ttl time.Duration, publicAddr string, clusterName string) ([]byte, error) {
+	user, err := a.Identity.GetUser(username, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	checker, err := services.FetchRoles(user.GetRoles(), a.Access, user.GetTraits())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certs, err := a.generateUserCert(certRequest{
+		user:      user,
+		publicKey: publicKey,
+		checker:   checker,
+		ttl:       ttl,
+		// Set the login to be a random string. Application certificates are never
+		// used to log into servers but SSH certificate generation code requires a
+		// principal be in the certificate.
+		traits: wrappers.Traits(map[string][]string{
+			teleport.TraitLogins: []string{uuid.New()},
+		}),
```

Comment:
I think the fact that this is a UUID is causing some log spam:

```
Oct 29 22:51:54 hades teleport[3166689]: DEBU             Skipping login be605d4e-2799-4769-ba0e-fe9c6b8861c1, not a valid Unix login. services/role.go:316
Oct 29 22:51:54 hades teleport[3166689]: DEBU             Skipping login be605d4e-2799-4769-ba0e-fe9c6b8861c1, not a valid Unix login. services/role.go:316
Oct 29 22:51:54 hades teleport[3166689]: DEBU             Skipping login be605d4e-2799-4769-ba0e-fe9c6b8861c1, not a valid Unix login. services/role.go:316
Oct 29 22:51:54 hades teleport[3166689]: INFO             Round trip: POST /json, code: 200, duration: 20.851603ms tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:51:55 hades teleport[3166689]: INFO             Round trip: POST /json, code: 200, duration: 193.962724ms tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:51:59 hades teleport[3166689]: INFO             Round trip: POST /json, code: 200, duration: 5.120823111s tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:52:08 hades teleport[3166689]: DEBU             Skipping login be605d4e-2799-4769-ba0e-fe9c6b8861c1, not a valid Unix login. services/role.go:316
Oct 29 22:52:09 hades teleport[3166689]: DEBU             Skipping login be605d4e-2799-4769-ba0e-fe9c6b8861c1, not a valid Unix login. services/role.go:316
Oct 29 22:52:09 hades teleport[3166689]: DEBU             Skipping login be605d4e-2799-4769-ba0e-fe9c6b8861c1, not a valid Unix login. services/role.go:316
Oct 29 22:52:09 hades teleport[3166689]: INFO             Round trip: POST /json, code: 200, duration: 17.994725ms tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:52:10 hades teleport[3166689]: INFO             Round trip: POST /json, code: 200, duration: 189.447109ms tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:52:13 hades teleport[3166689]: INFO             Round trip: POST /json, code: 200, duration: 5.124453842s tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:52:15 hades teleport[3166689]: DEBU [PROXY:AGE] Ping -> 142.167.142.235:3080. leaseID:1 target:teleport.webvict.im:3080 reversetunnel/agent.go:405
Oct 29 22:52:19 hades teleport[3166689]: ERRO             Error forwarding to /signalr/reconnect?transport=longPolling&connectionToken=PlR1wsfNmfuMNn4iqR8JFibjwopBg5IsoqqQKsmzQHIpIoWsoMW%2FrNnmmHuYrRhnNLQQqScdrEaHMu%2By4x8E%2BCzwCuFPoa2lAOzLtfNkqoDtLNE%2Fo73WiDpcy4h%2FNe0dDAvOG8BH9O6rwIcIp%2FaJJCHkMZzg4CpxuvZk0PGZE%2F0LV8iA2FG1yl0%2FHZiIUhX1zUPVmGlNGRhIxXSGt1j0V5wo3cNALBubxkeoGJ0vzlW4iJSK7fyXefjJ7F6d410ChC2n8CuCKUBgCzJjligxzBETBfdeCJVMracB9kc2NNC5VwhPF2ImNbveZV3ZtrSV&messageId=d-C994E491-I%2C44&apiKey=ea59011c5501420780aee7beb5930dc8&tid=7&_=1604021796102, err: writer is closed forward/fwd.go:180
Oct 29 22:52:19 hades teleport[3166689]: DEBU             Skipping login 5a3729a7-23f1-4ecc-b2d6-5894d0a0cc70, not a valid Unix login. services/role.go:316
Oct 29 22:52:19 hades teleport[3166689]: INFO             Round trip: GET /signalr/ping?apiKey=ea59011c5501420780aee7beb5930dc8&_=1604021796103, code: 200, duration: 4.101691ms tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:74656c65706f72742e776562766963742e696d.teleport.cluster.local forward/fwd.go:196
Oct 29 22:52:19 hades teleport[3166689]: DEBU             Skipping login 5a3729a7-23f1-4ecc-b2d6-5894d0a0cc70, not a valid Unix login. services/role.go:316
```

It's 36 characters but the maximum allowed length is 32: https://github.com/gravitational/teleport/blob/e32db1d2f8e989a50251d5027e7a1e8120f12678/vendor/github.com/gravitational/configure/cstrings/user.go#L8-L35